compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,677 @@
+//! This module provides a framework on top of the normal MIR dataflow framework to simplify the
+//! implementation of analyses that track the values stored in places of interest.
+//!
+//! The default methods of [`ValueAnalysis`] (prefixed with `super_` instead of `handle_`)
+//! provide some behavior that should be valid for all abstract domains that are based only on the
+//! value stored in a certain place. On top of these default rules, an implementation should
+//! override some of the `handle_` methods. For an example, see `ConstAnalysis`.
+//!
+//! An implementation must also provide a [`Map`]. Before the anaylsis begins, all places that
+//! should be tracked during the analysis must be registered. The set of tracked places cannot be
+//! changed during the analysis.
+
+use std::fmt::{Debug, Formatter};
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_target::abi::VariantIdx;
+
+use crate::{
+    fmt::DebugWithContext, lattice::FlatSet, Analysis, AnalysisDomain, CallReturnPlaces,
+    JoinSemiLattice, SwitchIntEdgeEffects,
+};
+
+pub trait ValueAnalysis<'tcx> {
+    /// For each place of interest, the analysis tracks a value of the given type.
+    type Value: Clone + JoinSemiLattice + HasBottom + HasTop;
+
+    const NAME: &'static str;
+
+    fn map(&self) -> &Map;
+
+    fn handle_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        self.super_statement(statement, state)
+    }
+
+    fn super_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        match &statement.kind {
+            StatementKind::Assign(box (place, rvalue)) => {
+                self.handle_assign(*place, rvalue, state);
+            }
+            StatementKind::SetDiscriminant { .. } => {
+                // Could treat this as writing a constant to a pseudo-place.
+            }
+            StatementKind::CopyNonOverlapping(..) => {
+                // FIXME: What to do here?
+            }
+            StatementKind::StorageLive(..)
+            | StatementKind::StorageDead(..)
+            | StatementKind::Deinit(_) => {
+                // Could perhaps use these.
+            }
+            StatementKind::Nop
+            | StatementKind::Retag(..)
```

Comment:
Can you please help me understand what effect `Retag` would have regarding this analysis? As far as I understand, it does not influence the underlying values.

Reply:
It does. The Miri implementation can be found [here](https://github.com/rust-lang/miri/blob/284b59c4dc89ef7fe44ac0b372394081fcbb88f5/src/stacked_borrows/mod.rs#L899). The tldr though is that it literally reads the value that is currently found at the given place, computes a new value, and writes that new value to the place. The fact that the new value only differs from the old one by provenance is immaterial.

Reply:
I think what you are suggesting is that the concrete state should be based on Miri-like semantics (that include provenance), right? This actually makes a lot of sense, I like it a lot. However, I'm not sure about a default implementation like
```rust
self.handle_assign(*place, &Rvalue::Use(Operand::Move(*place)), state);
```
or something along those lines. If an analysis tracks abstract state related to provenance, then it must override the `Retag` statement anyway as the default implementation would be unsound. If it does not, then it doesn't need the default implementation. But maybe I'm wrong. What do you think?

Reply:
> I think what you are suggesting is that the concrete state should be based on Miri-like semantics (that include provenance), right?

Yes. To be clear, those are the only semantics Mir has. Provenance is not abstract state in the way that analysis results are; it is real state that happens to be optimized out when Mir is compiled to assembly.

> ```rust
> self.handle_assign(*place, &Rvalue::Use(Operand::Move(*place)), state);
> ```

This seems wrong. It fails to represent the mutation that happens. (also, self assignments like this are UB in Mir)

> Edit: Or maybe it depends on how conservative we want the default implementation to be. For example, with const prop, you can remove arbitrary handle_ overrides and the analysis stays sound. If we want this property in general, then I think we must flood places that are retagged because it changes the concrete state (only provenance, but that's enough).

Yes, exactly this. We don't know what the value at the place after the retag is, and so the analysis results should reflect that. If someone wants to use this framework in a situation where they do know what the result is (or can otherwise soundly approximate it), then they should override things as appropriate.

Reply:
> Yes, exactly this. We don't know what the value at the place after the retag is, and so the analysis results should reflect that. If someone wants to use this framework in a situation where they do know what the result is (or can otherwise soundly approximate it), then they should override things as appropriate.

I think we are on the same page now. I really like this direction.

What's below is something I have been thinking about. It's probably not complete. The code maybe doesn't compile. And I'm surely missing some things regarding aliasing. I will think about this some more, but here we go.

Let's assume an abstracte state $s \in \texttt{unreachable} \cup (\mathbb{P} \to \mathbb{V})$ at a given program point. The concretization $\gamma(s)$ of this state should probably be a subset of all concrete Miri states at this program point that are contained in UB-free executions. If $s = \texttt{unreachable}$, we define $\gamma(s) = \set{}$. Otherwise, $\gamma(s)$ contains all those Miri states from UB-free executions at this program point for which the following does not panic:

```rust
for (p, v) in s {
    let result = self.eval_place_to_op(p, None).and_then(|op| self.read_immediate(op));
    match result {
        Ok(imm_ty) => check(imm_ty, v),
        Err(_) => ()  // ignore
    }
}
```

And `check` is defined by the analysis. For instance, if we are doing constant propagation, then

```rust
fn check(imm_ty: ImmTy<'tcx, Provenance>, v: FlatSet<(ScalarInt, Ty<'tcx>)>) {
    match v {
        FlatSet::Top => (),
        FlatSet::Elem((scalar_int, ty)) => {
            assert!(imm_ty.layout.ty == ty);
            assert!(imm_ty.to_scalar().assert_int() == scalar_int);
        }
        FlatSet::Bottom => panic!(),
    } 
}
```
A hypothetical proof would have to show that $f(\gamma(s)) \subseteq \gamma(f'(s))$, where $f$ takes a set of Miri states and and takes one step for each of them, and $f'$ is a step in the dataflow analysis. This is of course for every pair $(f, f')$ and every $s$.

(Just to be clear: I'm not proposing actually doing the proof outlined above.)

This also opens the possibility for verifying that the analysis does the right thing for a concrete Miri execution starting with $x_0$ at the function entry point (assuming that the execution will not be UB). Start with the initial dataflow state $s_0$ and verify $x_0 \in \gamma(s_0)$ (you don't need to compute $\gamma$ to do that, just check whats above for $x_0$). Then take one Miri step with $f_0$ to get $x_1 = f_0(x_0)$. "Execute" the same step in the analysis: $s_1 = f_0'(s_0)$. Verify $x_1 \in \gamma(s_1)$. Continue like that, without performing any lattice joins. Not sure whether that is actually useful, but it seems quite nice.

What do you think? Would you want something other than `read_immediate(...)`?

Reply:
Sorry for the delayed reply, currently traveling.

> Should `ImmTy` be the concrete domain for a tracked place?

I don't think there's any need to restrict ourselves to `ImmTy`, we will probably want to extend this const-prop to non-scalars at some point. The right answer is I think a [`Value`](https://github.com/RalfJung/minirust/blob/master/lang/values.md) in the minirust sense. At least, I don't know any alternative under which a standard const-prop transform would continue to be sound.

Reply:
No worries, I'm also pretty busy at the moment, which is why there was no progress recently.

I think we have two options here:
1. We could focus first and foremost on getting a sound constant propagation without changing too much. This would mean that we can ignore some of the identified shortcomings for now.
2. We could focus on improving the surrounding framework, such that it handles retags, discriminants, etc. properly. This will take more time initially, but will pay off if we want to do have a more precise constant propagation or other analyses (and we probably do). But this requires a solid model of what we actually want to track.

Regarding `ImmTy`: I initially suggested it because I was looking for an "atomic" object, because I was not sure how to deal with overlap. Turns out that `ImmTy` is not a good choice for "atomic" objects (probably something like `Scalar` fits better), but overlapping objects might actually be not too hard. 

Do you have any concrete analyses in mind that would require tracking non-scalars in the abstract domain? Also, I'm not really sure whether I follow your last point (that there is no alternative besides MiniRusts `Value`). Would be great if you could help me understand what you mean by that. Thank you!

Reply:
Note that retags don't exist during normal builds, they are added only for Miri. So if there is an unsoundness caused by ignoring how retags mutate a place's content, then that unsoundness is still present in regular compilation where those retags are missing.  If retags inhibit optimizations, we need to carefully consider whether those optimizations don't have to also be inhibited during regular codegen where no retags exist.

Reply:
> But upon further though, it's actually not at all. We can continue to track only the "atomic" parts and then "re-aggregate" them upon seeing a use of the "outer" place.

Yes, and I think that this is also the cleanest way to represent this. 

> The right version is probably to use ImmTy, because we want to be able to store fat pointers, which cannot be stored in scalars.

Fat pointers are pairs of scalars, no? So they could also be represented in that way (although I think that such pairs are currently not representable, since they are not real places). I would postpone this, but I have already some ideas on how it can be done.

Reply:
Fat pointers have ScalarPair ABI, but I expect that in Minirust they will be "atomic" and not represented as an aggregate

Reply:
Ah, I see. In any case, I guess we better leave this for a future PR, right?

Reply:
Yes, absolutely. Thanks for working this out!